python/ray/tests/test_gcs_ha_e2e.py

```diff
@@ -187,6 +204,38 @@ def reconfig():
     assert output.exit_code == 0
 
 
+@pytest.mark.skipif(sys.platform != "linux", reason="Only works on linux.")
+def test_ray_nodes_liveness(docker_cluster):
+    get_nodes_script = """
+import ray
+ray.init("auto")
+print(sum([1 if n["Alive"] else 0 for n in ray.nodes()]))
+"""
+    header, worker = docker_cluster
+
+    def check_alive(n):
+        output = worker.exec_run(cmd=f"python -c '{get_nodes_script}'")
+        assert output.exit_code == 0
+        text = output.output.decode().strip().split("\n")[-1]
+        print("Alive nodes: ", text)
+        return n == int(text)
+
+    # Make sure two nodes are alive
+    wait_for_condition(check_alive, n=2)
+    print("header killed")
+    header.kill()
+
+    sleep(2)
+
+    header.restart()
+    # When GCS restarts, a new raylet is added
+    # and the old dead raylet is going to take a while to be marked dead.
+    # So there should be 3 alive nodes
+    wait_for_condition(check_alive, timeout=10, n=3)
```

Comment:
Oh, it won't We have health check implemented, the protocol is similar to k8s. Basically it will only be marked dead after x times failure. And we do check every y seconds.
https://github.com/ray-project/ray/blob/master/src/ray/common/ray_config_def.h#L738
So as long as this line is not too late (delayed for 15s maybe?), it's ok.